README.md
# shaderSpace

The Autodesk Maya script that can be build shader with texture and connections more faster and convenient.

# Version

1.2.0

# How to use

Copy the scripts and prefs folder into your Maya script path,  
Default path should be 

+ C:\%USERNAME%\document\maya\%MAYA_VERSION% (Windows)

Open Maya application, enter below commands into python command window :  

```python
import shaderSpace
reload(shaderSpace)
```

If success, It will lanuch shaderSpace.

# Maya Support

Maya 2014, 2015, 2016, 2016.5, 2017, 2018

# Renderer Support

+ Maya Software Renderer.
+ Mental Ray Renderer.
+ Arnold Renderer (mtoa 2 & 3).
+ VRay Renderer.

# OS Support

Windows 10


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## Unreleased

## [1.2.0] : 2018-08-15

+ Fix the problem that script can't open in Maya 2016.5.
+ Delete turtle ambient occlusion bake module.

## [1.1.0] : 2017-08-15

+ Remove "Baker" function.
+ Now working on Maya 2017 (PySide2).
+ Fix problem that the new cm on Maya 2015 oldest version.
+ Fix color-profile incorrect.

## [1.0.03] : 2017-07-07

+ Fix create texture color management node connected in Maya 2015 Extension 1 or upper versions.

## [1.0.02] : 2016-12-12

+ New feature release.

## [0.5.05] : 2016-07-24

+ Add color management profile for maya 2015.
+ Add save/ load settings and lock settings.
+ Fix something problems.

## [0.5.04] : 2016-07-09
    
+ New feature for shaderSpace tool, using PySide.

## [0.4.34] : 2016-05-26

+ Add log switcher.

## [0.4.33] : 2016-05-25

+ Fix bugs.
+ Add log window.

## [0.4.32] : 2016-05-22

+ Add copyright header.

## [0.4.31] : 2016-05-22

+ Optimize the code structure.

## [0.4.30] : 2016-05-21

+ Optimize the code structure.

## [0.4.20] : 2016-05-19

+ Add new way in place2d - shared place2dTexture and non-shared.

## [0.4.11] : 2016-05-18

+ Fix the problem that gamma correct can't turn off.
+ Refine some code.

## [0.4.01] : 2016-05-15

+ Fix the problem that whitespace of channel's name.
+ Add shader's attribute preset, Now can be set preset when creating.

## [0.3.6] : 2016-04-13

+ Add gamma correct method in VRay.

## [0.3.5] : 2016-05-29

+ Channels have been changed, increase to 12 channels.
+ Fix core process

## [0.3.4] : 2016-05-20

+ Change gamma correct rule.

## [0.3.3] : 2016-05-06

+ Fix arnold shader config.

## [0.3.2] : 2016-05-01

+ Add process window in the tools, and fix same UI problems.

## [0.3.1] : 2016-02-27

+ The method additional that can be rebuild shader connections to difference shader.
+ Add presets UI.

## [0.3.0] : 2016-02-23

+ Add function that can be checkout material's type before creation.
+ Main icon changed.

## [0.2.9] : 2016-02-19

+ Refine config options.
+ Fix default options.

## [0.2.8] : 2016-02-16

+ Fix problems of export mesh.
+ Fix relatives of material's channel connections.

## [0.2.7] : 2016-02-09

+ Refine some UI detail.
+ Fix save options problems.
+ Fix load options problems.

## [0.2.6] : 2016-02-08

+ load configuration function has been implemented.

## [0.2.5] : 2016-02-06

+ The function export mesh and uvsnapshot can be support sets now.
+ Fix some UI problems.

## [0.2.3] : 2016-02-06

+ Refine serval problems

## [0.2.2] : 2016-01-21

+ Add try & except block, try to fix some problems about optionVar.
+ Fix some tools bugs.

scripts/shaderSpace/base.py
'''
    Base UI Class for shaderSpace Rapid Shader Workflow Tool in Maya

    Made by Chia Xin Lin, Copyright (c) 2016 by Chia Xin Lin
    E-Mail : nnnight@gmail.com
    Github : http://github.com/chiaxin
'''
import maya.cmds as mc

class BaseUI:
    Win = 'BASE_UI'
    Frl = 'BASE_UI_FRL'
    width = 240
    height= 360

    def build(self, menu, blocks, uit, title):
        self.destroy()
        mc.window(self.Win, w=self.width, h=self.height, t=title)
        if uit is not None:
            uit.build()
            uit.push()
        if menu is not None: menu.build()
        self.Frl = mc.frameLayout(self.Frl)
        if blocks is not None and len(blocks):
            for block in blocks:
                block.build()
        mc.setParent('..')
        if uit is not None: uit.pop()

    def destroy(self):
        if self.exists():
            mc.deleteUI(self.Win)

    def exists(self):
        return mc.window(self.Win, q=True, ex=True)

    def show(self):
        if self.exists():
            mc.showWindow(self.Win)
            mc.window(self.Win, e=True, w=self.width, h=self.height)

class BaseUIT:
    Uit = 'BASE_UIT'

    def build(self):
        if mc.uiTemplate(self.Uit, q=True, ex=True):
            mc.deleteUI(self.Uit, uit=True)
        mc.uiTemplate(self.Uit)
        self.templates()

    def templates(self):
        # Control
        mc.button(dt=self.Uit, h=30)
        mc.text(dt=self.Uit, font='fixedWidthFont', w=100)
        mc.textField(dt=self.Uit, w=150, h=26)
        mc.textFieldGrp(dt=self.Uit, w=300, h=30, cw2=(80, 240))
        mc.textFieldButtonGrp(dt=self.Uit, w=320, cw3=(80, 240, 60), bl='browse')

        # Layout
        mc.frameLayout(dt=self.Uit, cl=False, cll=False, lv=False, mh=5, mw=5)
        mc.columnLayout(dt=self.Uit, adj=True , rs=3, cal='left')
        mc.radioButtonGrp(dt=self.Uit, cw4=(60, 60, 60, 60), cw3=(60, 60, 60))

    def push(self):
        mc.setUITemplate(self.Uit, pushTemplate=True)

    def pop(self):
        mc.setUITemplate(popTemplate=True)

class BaseBlock:
    Frl = 'BLOCK_FRL'
    Label = 'Base Block'
    width = 256
    height= 128

    def build(self):
        self.Frl = mc.frameLayout(self.Frl, l=self.Label, w=self.width, h=self.height)
        self.content()
        mc.setParent('..')

    def content(self): # This must to override
        pass


scripts/shaderSpace/tools.py
# -*- coding: utf-8 -*-
# Tools for shaderSpace
#
# Made by Chia Xin Lin, Copyright (c) 2016 by Chia Xin Lin
#
# E-Mail : nnnight@gmail.com
#
# Github : http://github.com/chiaxin
#
# Licensed under MIT: https://opensource.org/licenses/MIT
#
from os import listdir
from os.path import isfile, isdir, dirname, exists
from config import kChannelList
import maya.cmds as mc
import maya.mel as mel

def exportShaders(export_path, mode):
    if not export_path:
        mc.warning('Please enter a output folder!')
        return -1
    if not exists(export_path):
        mc.warning('Directory is not exists :'+export_path)
        return -1
    shading_groups = []
    if mode == 0:
        shading_groups = [s for s in mc.ls(type='shadingEngine')
            if s not in ('initialParticleSE', 'initialShadingGroup')]
    elif mode == 1:
        selections = mc.ls(sl=True)
        if not selections:
            mc.warning(
                'Please select at least one shader or mesh, nurbsShapes.')
            return -1
        for sel in selections:
            connected_sg = mc.listConnections(sel, s=False,
                d=True, type='shadingEngine')
            if not connected_sg:
                continue
            shading_groups.extend([s for s in connected_sg
                if s not in ('initialParticleSE', 'initialShadingGroup')])
        shading_groups = list(set(shading_groups))
    else:
        return -1
    if not shading_groups:
        mc.warning('There are no any shaders can be export!')
        return -1
    if not _userConfirm('Export Shaders', '\n'.join(shading_groups)):
        return -1
    connections = []
    fullpath = ''
    amout = 0
    process_max = len(shading_groups)
    mc.progressWindow(title='Export Shaders', progress=amout,
        status='Export start...', isInterruptable=True, max=process_max)
    for sg in shading_groups:
        if mc.progressWindow(q=True, isCancelled=True):
            break
        fullpath = '{0}{1}.ma'.format(export_path, sg.replace(':', '_'))
        connections = mc.listHistory(sg, allFuture=True, pruneDagObjects=True)
        mc.select(connections, replace=True, noExpand=True)
        try:
            mc.file(fullpath, force=True, exportSelected=True,
                type='mayaAscii', options='v=0;')
            amout += 1
            mc.progressWindow(e=True, progress=amout, 
                status='Export : {0} / {1}'.format(amout,  process_max))
        except:
            raise
        mc.select(cl=True)
    mc.progressWindow(endProgress=True)

def exportPolygons(export_path, export_type, exclude, include, source):
    if not export_path:
        mc.warning('Please enter a output folder!')
        return -1
    elif not exists(export_path):
        mc.warning('Directory is not exists : {0}'.format(export_path))
        return -1
    objExportPlugin = mc.pluginInfo('objExport', query=True, loaded=True)
    if export_type == 'OBJ' and not objExportPlugin:
        load_plugins = mc.loadPlugin('objExport')
        if load_plugins[0] != 'objExport':
            mc.warning('obj export plugin load failed')
            return -1
        else:
            print('Plugin loaded : obj export')
    export_info = {}
    if export_type == 'OBJ':
        export_info['typ'] = 'OBJexport'
        export_info['ext'] = 'obj'
        export_info['opt'] = \
            'group=0;ptgroups=0;materials=0;smoothing=1;normals=1'
    elif export_type == 'Maya Ascii':
        export_info['typ'] = 'mayaAscii'
        export_info['ext'] = 'ma'
        export_info['opt'] = 'v=0;'
    elif export_type == 'Maya Binary':
        export_info['typ'] = 'mayaBinary' 
        export_info['ext'] = 'mb'
        export_info['opt'] = 'v=0;'
    elif export_type == 'FBX':
        export_info['typ'] = 'fbx'
        export_info['ext'] = 'fbx'
        export_info['opt'] = ''
    else:
        mc.warning('Failed : Unknown export type : {0}'.format(export_type))
    pairs_mesh = {}
    if source == 0:
        pairs_mesh = _getPolygonFromDisplayLayers()
    elif source == 1:
        pairs_mesh = _getPolygonsFromSets()
    elif source == 2:
        selections = mc.ls(sl=True, l=True)
        if selections:
            pairs_mesh = {'export_mesh': selections}
        else:
            mc.warning('# Please select objects!')
            return -1
    #
    if not pairs_mesh:
        return -1
    if not _userConfirm('Export Meshes', '\n'.join(pairs_mesh.keys())):
        return -1
    store_selections = mc.ls(sl=True, l=True)
    amout = 0
    process_max = len(pairs_mesh)
    # Progress Window
    mc.progressWindow(title='Export Shaders', progress=amout,
        status='Export start...', isInterruptable=True, max=process_max)
    for key in pairs_mesh.keys():
        mc.select(cl=True)
        fullpath = '{0}/{1}.{2}'.format(export_path, key, export_info['ext'])
        for item in pairs_mesh[key]:
            shortName = item.split('|')[-1]
            if len(include) == 0 or shortName.find(include) >= 0 and \
            len(exclude) == 0 or shortName.find(exclude) == -1:
                mc.select(item, add=True)
        if len(mc.ls(sl=True)) == 0:
            continue
        try:
            mc.file(fullpath, f=True, exportSelected=True,
                type=export_info['typ'], options=export_info['opt'])
            amout += 1
            mc.progressWindow(e=True, progress=amout, 
                status='Export : {0} / {1}'.format(amout,  process_max))
        except:
            raise
        print('The meshes have been exported : '+fullpath)
    mc.progressWindow(endProgress = True)
    if objExportPlugin is not True:
        unloaded_plugins = mc.unloadPlugin('objExport')
        if unloaded_plugins[0] == 'objExport':
            print('Unload plug-in : objExport')
        else:
            mc.warning('Failed to unload plug-in : objExport')
    if store_selections:
        mc.select(store_selections, r=True)
    else:
        mc.select(cl=True)

def uvSnapshot(export_path, ext, res, colorSymbol, source=0):
    if not export_path:
        mc.warning('Please enter a output folder!')
        return -1
    if not isdir(export_path):
        mc.warning('The directory is invalid : {0}'.format(export_path))
        return -1
    pairs_mesh = {}
    # Get source 0: from display layer, 1:from sets, 2:from selected
    if source == 0:
        pairs_mesh = _getPolygonFromDisplayLayers()
    elif source == 1:
        pairs_mesh = _getPolygonsFromSets()
    elif source == 2:
        selections = mc.ls(sl=True, l=True)
        if selections:
            pairs_mesh = {'uv_snapshot': selections}
        else:
            mc.warning('# Please select objects!')
            return -1
    if not pairs_mesh:
        mc.warning('No any meshs can be snapshot')
        return -1
    if not _userConfirm('UV Snapshot', '\n'.join(pairs_mesh.keys())):
        return
    # Store previous selections
    store_selections = mc.ls(sl=True)
    amout = 0
    process_max = len(pairs_mesh)
    # UV Snapshot Color
    R, G, B = 0, 1, 2
    if colorSymbol == 'Black':
        color = (0, 0, 0)
    elif colorSymbol == 'White':
        color = (255, 255, 255)
    elif colorSymbol == 'Red':
        color = (255, 0, 0)
    else:
        color = (255, 255, 255)
    # Progress Window
    mc.progressWindow(title='UV Snapshot', progress=amout,
        status='Snapshot start...', isInterruptable=True, max=process_max)
    #
    for key in pairs_mesh.keys():
        mc.select(cl=True)
        fullpath = '{0}/{1}.{2}'.format(export_path, key , ext)
        try:
            mc.select(pairs_mesh[key], r=True)
            mc.uvSnapshot(overwrite=True,
                          redColor=color[R],
                          greenColor=color[G],
                          blueColor=color[B],
                          xResolution=int(res),
                          yResolution=int(res),
                          fileFormat=ext,
                          antiAliased=True,
                          name=fullpath)
            amout += 1
            mc.progressWindow(e=True,
                              progress=amout,
                              status='Snapshot : {0} / {1}'.format(
                              amout,  process_max))
        except:
            mc.progressWindow(endProgress=True)
            raise
        print('UV Snapshot > '+fullpath)
    # End the progress window
    mc.progressWindow(endProgress=True)
    if store_selections:
        mc.select(store_selections, r=True)
    else:
        mc.select(cl=True)

def createPhotoshopFile(export_path, uv_path, channels, res):
    if not export_path:
        mc.warning('# Please enter a output folder!')
        return -1
    if exists(export_path):
        mc.warning('# File is exists, Can not overwrite : '+export_path)
        return -1
    if not _userConfirm('Photoshop file', 'Create {0} ?'.format(export_path)):
        return -1
    command = 'psdTextureFile -uvt true -psf \"{0}\" '.format(export_path)
    command += '-xr {0} -yr {1} '.format(str(res), str(res))
    if isfile(uv_path):
        command += '-ssi \"{0}\" '.format(uv_path)
        print('# UV image found :' + uv_path)
    else:
        print('# UV image file was not found.')
    # Add layers flags
    channels_line = ''
    for idx, ch in enumerate(channels):
        #channels_line += '-chs {0} {1} {2} '.format(ch, str(idx), 'true')
        channels_line += '-chc {0} {1} 128 128 128 '.format(ch, str(idx))
    if channels_line:
        command += channels_line
    try:
        mel.eval(command)
    except:
        print command
        raise
    else:
        print '# Photoshop file create successiful! : ', export_path

def _userConfirm(title, message):
    ans = mc.confirmDialog(t=title, m=message, b=('Yes','No'), 
        db='Yes', cb='No', ds='No')
    if ans == 'Yes':
        return True
    return False

def _getPolygonFromDisplayLayers():
    visible_displaylayers = [ d for d in mc.ls(type='displayLayer')
        if mc.getAttr(d+'.visibility') and d not in ['defaultLayer']]
    if not visible_displaylayers:
        return {}
    polygonCollections = {}
    for dl in visible_displaylayers:
        itemsInLayer = mc.editDisplayLayerMembers(dl, q=True, fn=True)
        if not itemsInLayer:
            continue
        items = [o for o in itemsInLayer if mc.nodeType(o) in
            ['mesh', 'transform']]
        if not items:
            continue
        collections = []
        for item in items:
            if mc.nodeType(item) == 'mesh':
                collections.append(item)
            elif mc.nodeType(item) == 'transform':
                relatives = mc.listRelatives(item, ad=True,
                    type='mesh', ni=True, f=True)
                collections.extend(relatives)
        if collections:
            polygonCollections[dl] = collections
    return polygonCollections

def _getPolygonsFromSets():
    all_sets = [s for s in mc.ls(type='objectSet')
    if mc.nodeType(s) == 'objectSet' \
    and s not in ['defaultLightSet', 'defaultObjectSet']]
    if not all_sets:
        return {}
    polygonCollections = {}
    for se in all_sets:
        collections = []
        itemsInSet = [o for o in mc.sets(se, q=True) 
            if mc.nodeType(o) in ['mesh', 'transform']]
        if not itemsInSet:
            continue
        for item in itemsInSet:
            if mc.nodeType(item) == 'mesh':
                collections.append(item)
            elif mc.nodeType(item) == 'transform':
                relatives = mc.listRelatives(item, ad=True,
                    type='mesh', ni=True, f=True)
                collections.extend(relatives)
        if collections:
            polygonCollections[se] = collections
    return polygonCollections


scripts/shaderSpace/core.py
# -*- coding: utf-8 -*-
# shaderSpace core functions
#
# Made by Chia Xin Lin, Copyright (c) 2016 by Chia Xin Lin
#
# E-Mail : nnnight@gmail.com
#
# Github : http://github.com/chiaxin
#
# Licensed under MIT: https://opensource.org/licenses/MIT
#
import re
import sys
import os.path
from functools import partial
import maya.cmds as mc
import maya.mel as mel
import logging
from config import *

logging.basicConfig(levels=logging.DEBUG, format='%(levelname)s : %(message)s')

def isVaildName(name):
    matcher = re.match(r'[a-zA-Z_]\w+', name)
    if matcher:
        return True
    return False

def _substituteVariables(rule, **kwargs):
    # kwargs: asset, shader, user, version, channel
    if not rule:
        return ''
    # There get variable keywords and replace.
    rule = rule.replace('<asset>',  kwargs.get('asset', 'asset'))
    rule = rule.replace('<shader>', kwargs.get('shader', 'shader'))
    rule = rule.replace('<user>',   kwargs.get('user', 'user'))
    rule = rule.replace('<version>',kwargs.get('version', 'version'))
    rule = rule.replace('<root>',  mc.workspace(q=True, rootDirectory=True))
    # If channel is filled, replace it
    if kwargs.has_key('channel'):
        rule = rule.replace('<channel>', kwargs.get('channel', ''))
    # Determine camel case format
    buffer = rule.split('<c>')
    for idx in range(len(buffer)-1):
        if buffer[idx][-1].islower():
            buffer[idx+1] = buffer[idx+1][0].upper() + buffer[idx+1][1:]
        elif buffer[idx][-1].isupper():
            buffer[idx+1] = buffer[idx+1][0].lower() + buffer[idx+1][1:]
    return ''.join(buffer).replace('\\', '/')

def _connectToPlace2dTexture(fileNode, p2dNode):
    pairs = (('coverage', 'coverage'),
             ('translateFrame', 'translateFrame'),
             ('rotateFrame', 'rotateFrame'),
             ('mirrorU', 'mirrorU'),
             ('mirrorV', 'mirrorV'),
             ('stagger', 'stagger'),
             ('wrapU', 'wrapU'),
             ('wrapV', 'wrapV'),
             ('repeatUV', 'repeatUV'),
             ('offset', 'offset'),
             ('rotateUV', 'rotateUV'),
             ('noiseUV', 'noiseUV'),
             ('vertexUvOne', 'vertexUvOne'),
             ('vertexUvTwo', 'vertexUvTwo'),
             ('vertexUvThree', 'vertexUvThree'),
             ('vertexCameraOne', 'vertexCameraOne'),
             ('outUV', 'uv'),
             ('outUvFilterSize', 'uvFilterSize'))
    for attr in pairs:
        mc.connectAttr((p2dNode+'.'+attr[0]),
            (fileNode+'.'+attr[1]), force=True)

def _buildPlace2dTexture(mainname, mirrorU, mirrorV):
    p2d = mc.shadingNode('place2dTexture', name=mainname, asUtility=True)
    mc.setAttr(p2d+'.mirrorU', mirrorU)
    mc.setAttr(p2d+'.mirrorV', mirrorV)
    return p2d

def _isExistsNodeType(typ):
    return (typ in mc.ls(nodeTypes=True))

def createShader(*args, **kwargs):
    try:
        shaderType = kShadersList[kwargs['shaderSpaceShaderIntOptVar']]
        preset = kwargs['preset']
        cnames = kwargs['shaderSpaceAbbreviationsStrOptVars']
        checks = kwargs['shaderSpaceChannelsIntOptVars']
        filterTypeIdx = kwargs['shaderSpaceFilterTypeIntOptVar']
        autoPathRule= kwargs['shaderSpaceAutoPathRuleStrOptVar']
        nodeRules = kwargs['shaderSpaceNodeNameRuleStrOptVars']
        shaderRule = nodeRules[0]
        shadingRule = nodeRules[1]
        textureRule = nodeRules[2]
        bump2dRule = nodeRules[3]
        place2dRule = nodeRules[4]
        matinfoRule = nodeRules[5]
        options = kwargs['shaderSpaceOptionIntOptVars']
        assignToSelected = options[0]
        gammaCorrentMode = options[1]
        autoPathMode  = options[2]
        mirrorUvMode = options[3]
        sharedP2dMode = options[4]
        bumpValue = kwargs['shaderSpaceBumpValueFloatOptVar']
        replacements = kwargs['shaderSpaceNameStrOptVars']
        isAlphaIsLum = kwargs['shaderSpaceAlphaIsLuminanceIntOptVar']
        colorProfiles = kwargs['shaderSpaceColorProfileNameStrOptVars']
    except:
        raise
    # Check empty
    if not all(nodeRules):
        mc.warning('Maya nodes can not be empty')
        return '', ''
    elif not all(cnames):
        mc.warning('Channel abbreviations can not be empty')
        return '', ''
    elif not all(replacements):
        mc.warning('Name field can not be empty')
        return '', ''
    pSubstituteFunction = partial(
        _substituteVariables,
        asset=replacements[0],
        shader=replacements[1],
        user=replacements[2],
        version=replacements[3])
    if not _isExistsNodeType(shaderType):
        mc.warning(
            'Shader type is not exists. please check plug-in was appended')
        return '', ''
    # Shader name
    mainShader = pSubstituteFunction(shaderRule)
    # Save current selections
    previousSelected = mc.ls(
        sl=True, type=('mesh', 'nurbsSurface', 'transform'))
    # Shader name is invaild?
    if not isVaildName(mainShader):
        mc.warning('Invalid node name : '+mainShader)
        return '', ''
    # Is shader exists? if so, ask user.
    if mc.objExists(mainShader) and mc.confirmDialog(
        t='Exists', m='Shader is exists, continue?', button=['Yes', 'No'],
        db='Yes', cb='No', ds='No') == 'No': return '', ''
    # Shader create
    mainShader = mc.shadingNode(shaderType, name=mainShader, asShader=True)
    # Load preset, using source command
    if preset and preset != 'No Preset':
        presetPath = '{0}attrPresets/{1}/{2}.mel'.format(
            mc.internalVar(ups=True), shaderType, preset)
        if os.path.isfile(presetPath):
            mel.eval(r'applyPresetToNode "{0}" "" "" "{1}" 1'.format(
                mainShader, preset))
            print '{0} has been set : {1}'.format(mainShader, preset)
        else:
            mc.warning('Preset not found : {1}'.format(presetPath))
    # Shading Group create
    shadingGroup = mc.sets(renderable=True, noSurfaceShader=True,
        empty=True, name=pSubstituteFunction(shadingRule))
    # Rename material info
    materialInfo = mc.listConnections(shadingGroup,
        s=False, d=True, type='materialInfo')[0]
    materialInfo = mc.rename(materialInfo, pSubstituteFunction(matinfoRule))
    # Get connection pair
    connectTable = kRelatives[shaderType]
    # Connect shader to shading group, use different way
    if shaderType in kGeneralShadingGroupConnectMethod:
        _connectGeneralShadingGroup(mainShader, shadingGroup)
    elif shaderType in kMentalRayShadingGroupConnectMethod:
        _connectMentalRayShadingGroup(mainShader, shadingGroup)
    # Create and connect each channels
    newTextureList = []
    newTextureChannels = []
    # 2017-07-04 : After Maya 2015 extension 1, create texture(file)
    # need to use isColorManaged to plug color-management node.
    # mc.about(iv=True) will return a string such as
    # Autodesk Maya 2015 Extension 1 + SP6
    # if upper than 2015 extension 1, add isColorManaged argument
    create_texture_cmd = 'mc.shadingNode(\'file\', asTexture=True,'
    install_version_string = mc.about(iv=True)
    version_matcher = re.search(r'\d\d\d\d', install_version_string)
    if version_matcher:
        if version_matcher.group(0) == '2015'\
        and install_version_string.find('Extension') > 0:
            create_texture_cmd += 'isColorManaged=True,'
        elif int(version_matcher.group(0)) > 2015:
            create_texture_cmd += 'isColorManaged=True,'
    for idx, pairs in enumerate(connectTable):
        if not checks[idx]:
            continue
        # Texture node create
        #filenode = mc.shadingNode('file',
        #    name=pSubstituteFunction(textureRule, channel=cnames[idx]),
        #        asTexture=True)
        create_texture_temp_cmd = create_texture_cmd + (
            'name=\"' + pSubstituteFunction(
                textureRule, channel=cnames[idx]) + '\")'
        )
        filenode = eval(create_texture_temp_cmd)
        # aAttr is output attribute from texture,
        # bAttr is input attribute into shader.
        aAttr = pairs[0]
        bAttr = pairs[1]
        # This texture is a color map?
        withColorChannel = (aAttr == 'outColor')
        # Exclude bump map in VRay, because Bump map in VRay is outColor way.
        if shaderType in kVrayColorMangementShaders and bAttr == 'bumpMap':
            withColorChannel = False
        # Gamma correct setting :
        # In maya built-in shader or MentalRay, 
        # We need to set linear profile if is not color channel.
        # In VRay, we need to add VRay degamma attribute, 
        # And set it to sRGB if is color channel.
        # 2017/08/14 Fix :
        cm_is_exists = mel.eval("int $int =`exists colorManagementPrefs`") != 0
        if gammaCorrentMode:
            if shaderType in kColorManagementShaders and not withColorChannel:
                if not cm_is_exists:
                    _setColorProfileOld(filenode, 2)
                else:
                    _setColorProfileNew(filenode, colorProfiles[1])
            elif shaderType in kColorManagementShaders and withColorChannel:
                if not cm_is_exists:
                    # Do nothing
                    pass
                else:
                    _setColorProfileNew(filenode, colorProfiles[0])
            elif shaderType in kVrayColorMangementShaders and withColorChannel:
                _setVrayFileGamma(filenode,
                    kVrayDegammaMethod, kVrayDegammaValue)
        # If auto file path is 0(auto fill) or 1(auto fill is exists)
        if autoPathMode in (0, 1):
            fileTextureName = pSubstituteFunction(
                autoPathRule, channel=cnames[idx])
            if autoPathMode == 1 and not os.path.isfile(fileTextureName):
                # Skip not exists
                pass
            else:
                # Fill texture name
                mc.setAttr(filenode + '.fileTextureName',
                    fileTextureName, type='string')
        # If texture is outAlpha and AlphaIsLum mode is on
        if aAttr == 'outAlpha' and isAlphaIsLum:
            mc.setAttr(filenode+'.alphaIsLuminance', 1)
        # Set file's filter
        mc.setAttr(filenode+'.filterType', filterTypeIdx)
        # Here is deal with bump texture, index(9)
        if idx == 9:
            # Maya standard bump connect way
            if shaderType in kGeneralBumpMapMethod:
                _buildGeneralBumpMap(
                                    filenode+'.outAlpha',
                                    mainShader+'.normalCamera',
                                    pSubstituteFunction(bump2dRule), 
                                    bumpValue)
            # Mental ray bump connect way
            elif shaderType in kMentalRayBumpMapMethod:
                _buildGeneralBumpMap(
                                    filenode+'.outAlpha',
                                    mainShader+'.'+kMentalRayBumpMapChannel, 
                                    pSubstituteFunction(bump2dRule), 
                                    bumpValue)
            # Vray bump connect way
            elif shaderType in kVrayBumpMapMethod:
                _buildVrayBumpMap(
                                filenode+'.outColor',
                                mainShader+'.bumpMap',
                                bumpValue)
        else:
            mc.connectAttr(filenode+'.'+aAttr, mainShader+'.'+bAttr)
        newTextureList.append(filenode)
        newTextureChannels.append(cnames[idx])
    uvMirrorMethods = ((0, 0), (1,0), (0,1), (1,1))
    uvMirror = uvMirrorMethods[mirrorUvMode]
    # If Shared Place2dTexture is on,
    # we create a place2dTexture node connect to each map.
    place2dTextureNodeName = pSubstituteFunction(place2dRule)
    if sharedP2dMode == 0 and newTextureList:
        p2d = _buildPlace2dTexture(place2dTextureNodeName,
            uvMirror[0], uvMirror[1])
        for f in newTextureList:
            _connectToPlace2dTexture(f, p2d)
    # Otherwise, We need to create place2dTexture for each map.
    else:
        for idx, f in enumerate(newTextureList):
            # If user not provide <channel> flag,
            # It will suffix by alphabetic.
            if place2dRule.find('<channel>') == -1:
                p2d = _buildPlace2dTexture(
                    place2dTextureNodeName+'_'+chr(idx+65),
                        uvMirror[0], uvMirror[1])
            else:
                place2dTextureNodeName = pSubstituteFunction(
                    place2dRule, channel=newTextureChannels[idx])
                p2d = _buildPlace2dTexture(place2dTextureNodeName,
                    uvMirror[0], uvMirror[1])
            _connectToPlace2dTexture(f, p2d)
    # Assign to objects
    if assignToSelected == 1 and previousSelected:
        mc.select(previousSelected, r=True)
        mc.sets(e=True, fe=shadingGroup)
    mc.select(mainShader, r=True)
    return mainShader, shadingGroup

def _setColorProfileOld(filenode, colorProfile):
    # 0: Use Default Input Profile, 2: Linear sRGB, 3: sRGB
    mc.setAttr(filenode+'.colorProfile', colorProfile)

def _setColorProfileNew(filenode, colorProfile):
    logging.debug('%s : color-profile set %s'%(filenode, colorProfile))
    if colorProfile in \
    mc.colorManagementCatalog(ltc=True, type='input'):
        mc.setAttr(filenode+'.colorSpace', colorProfile, typ='string')
    else:
        mc.warning(
            'The {0} is not in color transforms'.format(colorProfile))

def _setVrayFileGamma(filenode, colorspace, gamma):
    try:
        mel.eval('vray addAttributesFromGroup '+filenode+' vray_file_gamma 1;')
        mc.setAttr(filenode+'.vrayFileGammaEnable', True)
        mc.setAttr(filenode+'.vrayFileColorSpace', colorspace)
        mc.setAttr(filenode+'.vrayFileGammaValue', gamma)
    except:
        mc.warning('Failed to set {0} to vray file gamma'.format(filennode))
        raise

def _buildGeneralBumpMap(output_map, input_shader, bump2dName, bumpValue):
    bump2d = mc.shadingNode('bump2d', name=bump2dName, asUtility=True)
    mc.connectAttr(output_map, bump2d+'.bumpValue', f=True)
    mc.connectAttr(bump2d+'.outNormal', input_shader, f=True)
    mc.setAttr(bump2d+'.bumpDepth', bumpValue)
    # Set attribute alphaIsLuminance checked in bump map
    mc.setAttr(output_map.split('.')[0]+'.alphaIsLuminance', 1)

def _buildVrayBumpMap(output_map, input_shader, bumpValue):
    mc.connectAttr(output_map, input_shader, f=True)
    mc.setAttr(input_shader.split('.')[0]+'.bumpMult', bumpValue)

def _connectGeneralShadingGroup(shader, shadingGroup):
    mc.connectAttr(shader+'.outColor', shadingGroup+'.surfaceShader', f=True)

def _connectMentalRayShadingGroup(shader, shadingGroup):
    for attr in ('miMaterialShader', 'miShadowShader', 'miPhotonShader'):
        mc.connectAttr(shader+'.message', shadingGroup+'.'+attr, f=True)
